core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -380,43 +370,48 @@ private void logClusterHealthStateChange(ClusterStateHealth previousStateHealth,
         }
     }
 
-    private boolean reroute(RoutingAllocation allocation) {
-        assert deassociateDeadNodes(allocation) == false : "dead nodes should be explicitly cleaned up. See deassociateDeadNodes";
+    private boolean hasDeadNodes(RoutingAllocation allocation) {
+        for (RoutingNode routingNode : allocation.routingNodes()) {
+            if (allocation.nodes().getDataNodes().containsKey(routingNode.nodeId()) == false) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void reroute(RoutingAllocation allocation) {
+        assert hasDeadNodes(allocation) == false : "dead nodes should be explicitly cleaned up. See deassociateDeadNodes";
 
-        boolean changed = false;
         // now allocate all the unassigned to available nodes
         if (allocation.routingNodes().unassigned().size() > 0) {
-            changed |= removeDelayMarkers(allocation);
-            changed |= gatewayAllocator.allocateUnassigned(allocation);
+            removeDelayMarkers(allocation);
+            gatewayAllocator.allocateUnassigned(allocation);
         }
 
-        changed |= shardsAllocator.allocate(allocation);
+        shardsAllocator.allocate(allocation);
         assert RoutingNodes.assertShardStats(allocation.routingNodes());
-        return changed;
     }
 
-    private boolean deassociateDeadNodes(RoutingAllocation allocation) {
-        boolean changed = false;
+    private void deassociateDeadNodes(RoutingAllocation allocation) {
         for (Iterator<RoutingNode> it = allocation.routingNodes().mutableIterator(); it.hasNext(); ) {
             RoutingNode node = it.next();
             if (allocation.nodes().getDataNodes().containsKey(node.nodeId())) {
                 // its a live node, continue
                 continue;
             }
-            changed = true;
+            allocation.changes().setChanged();
```

Comment:
Does this need to be explicitly set to changed if `node.copyShards()` returns no shards? Shouldn't the `failShard` call below handle setting the 'changed' flag on the allocation changes without this explicit change?